api/lib/opentelemetry/distributed_context/propagation/trace_parent.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+module OpenTelemetry
+  module DistributedContext
+    module Propagation
+      # A TraceParent is an implemenation of the W3C trace context specification
+      # {SpanContext}
+      class TraceParent
+        InvalidFormatError = Class.new(OpenTelemetry::Error)
+        InvalidVersionError = Class.new(OpenTelemetry::Error)
+        InvalidTraceIDError = Class.new(OpenTelemetry::Error)
+        InvalidSpanIDError = Class.new(OpenTelemetry::Error)
+
+        TRACE_PARENT_HEADER = 'traceparent'
+        SUPPORTED_VERSION = 0
+        private_constant :SUPPORTED_VERSION
+        MAX_VERSION = 254
+        private_constant :MAX_VERSION
+
+        REGEXP = /^([A-Fa-f0-9]{2})-([A-Fa-f0-9]{32})-([A-Fa-f0-9]{16})-([A-Fa-f0-9]{2})(-.*)?$/.freeze
+        private_constant :REGEXP
+
+        attr_reader :version, :trace_id, :span_id, :flags
+
+        def sampled?
+          flags.sampled?
+        end
+
+        def to_s
+          "00-#{trace_id}-#{span_id}-#{flags}"
+        end
+
+        def initialize(trace_id: nil, span_id: nil, version: SUPPORTED_VERSION, flags: OpenTelemetry::Trace::TraceFlags::DEFAULT)
+          @trace_id = trace_id
+          @span_id = span_id
+          @version = version
+          @flags = flags
+          raise ArgumentError, 'flags must be a TraceFlags' unless flags.is_a?(OpenTelemetry::Trace::TraceFlags)
+        end
+
+        def self.from_context(ctx)
+          TraceParent.new(trace_id: ctx.trace_id, span_id: ctx.span_id, flags: ctx.trace_flags)
+        end
+
+        def self.from_string(string)
+          TraceParent.new.parse(string)
+        end
+
+        def parse(string)
+          matches = match_input(string)
+
+          @version = parse_version(matches[1])
+          raise InvalidFormatError if version > SUPPORTED_VERSION && string.length < 55
+
+          @trace_id = parse_trace_id(matches[2])
+          @span_id = parse_span_id(matches[3])
+          @flags = parse_flags(matches[4])
+
+          self
+        end
+
+        private
+
+        def match_input(string)
+          matches = REGEXP.match(string)
+          raise InvalidFormatError, 'regexp match failed' if !matches || matches.length < 6
+
+          matches
+        end
+
+        def parse_version(string)
+          v = string.to_i(16)
+          raise InvalidFormatError, string unless v
+
+          raise InvalidVersionError, v if v > MAX_VERSION
+
+          v
+        end
+
+        def parse_trace_id(string)
+          raise InvalidTraceIDError, string if string == OpenTelemetry::Trace::INVALID_TRACE_ID
+
+          string.downcase
+        end
+
+        def parse_span_id(string)
+          raise InvalidSpanIDError, string if string == OpenTelemetry::Trace::INVALID_SPAN_ID
+
+          string.downcase
+        end
+
+        def parse_flags(string)
+          OpenTelemetry::Trace::TraceFlags.from_byte(string.to_i(16))
```

Comment:
D'you think we should add a `from_string` to `TraceFlags` and push the responsibility there?

Reply:
does the opentelemetry spec list the number of bytes and the hex format?  ie 2 digit hex.  
if it lives in the w3c but not in the OT spec I would say that would govern where it should live.

Reply:
The OT spec says TraceFlags is a single byte of boolean flags. Otherwise it defers to the W3C spec.

Reply:
then it should live here, as this is using a hex encoding of a byte.  Therefore the at the boundary it should be a native byte.